fission-codes/fission-cli
# Wrap existing haskell cli

# Description

This PR implements initial CLI functionality by wrapping the existing Haskell CLI binary. Following this PR, we will implement each command natively in Rust. 🦀 

This PR implements the the following commands:

- [x] `fission setup`
- [x] `fission user whoami` (and a `fission whoami` alias)
- [x] `fission user login`
- [x] `fission app register`
- [x] `fission app info`
- [x] `fission app publish`
- [x] `fission app delegate`

It also implements the following global flags:

- [x] `--verbose`
- [x] `--remote`
- [x] `--help` (descriptions in attributes per command)

We also started to add a `src/utils.rs` module, but it is not used in the wrapped CLI implementation.

## Link to issue

Closes #6

## Type of change

Please delete options that are not relevant.

- [x] Comments have been added

## Testing

Testing will be a bit manual, going through each command to see that it works. For each command, try some of the args and flags to see they work.

The recommended sequence is:

- Create a new user with `cargo run setup`
- Check the user is set up with `cargo run user whoami` and `cargo run whoami`
- Link the user to a browser with `cargo run user login`
- Register an app with `<path-to-binary> app register`
- Publish the app with `<path-to-binary> app publish`
- Get info on the app with `<path-to-binary> app info`
- Delegate access to the app with `<path-to-binary> app delegate`

The last `app` commands are best tested outside of the project directory, which is why`<path-to-binary>` is suggested for them. A temp directory with a text file can be created as a simple app to publish.

The `app delegate` command is slightly more involved than the other commands. At minimum, an app name and an audience must be specified. The app that was created and published while testing `app register` and `app publish` can be used for app name. The DID can be any valid `did:key`. For example:

```
<path-to-binary> app delegate -a short-old-tin-hero -d did:key:z6Mko4wBW851vmaB2VoA3dR6fisKK3Wn5bEWdhwnCw4yrRAm
```

The global `verbose` and `remote` commands can also be tested. Verbose is a bit easier and can be added along the way. Testing remote is essentially a separate pass through each of the commands.
